core/src/identity/rsa.rs

```diff
@@ -123,93 +123,148 @@ impl fmt::Debug for PublicKey {
 // Primer: http://luca.ntop.org/Teaching/Appunti/asn1.html
 // Playground: https://lapo.it/asn1js/
 
-lazy_static! {
-    /// The DER encoding of the object identifier (OID) 'rsaEncryption' for
-    /// RSA public keys defined for X.509 in [RFC-3279] and used in
-    /// SubjectPublicKeyInfo structures defined in [RFC-5280].
-    ///
-    /// [RFC-3279]: https://tools.ietf.org/html/rfc3279#section-2.3.1
-    /// [RFC-5280]: https://tools.ietf.org/html/rfc5280#section-4.1
-    static ref OID_RSA_ENCRYPTION_DER: DerObject =
-        DerObject {
-            tag: DerTag::x06,
-            value: DerValue {
-                data: vec![ 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01 ]
-            }
-        };
-}
+const ASN1_OBJECT_IDENTIFIER_TAG: u8 = 6;
+/// The DER encoding of the object identifier (OID) 'rsaEncryption' for
+/// RSA public keys defined for X.509 in [RFC-3279] and used in
+/// SubjectPublicKeyInfo structures defined in [RFC-5280].
+///
+/// [RFC-3279]: https://tools.ietf.org/html/rfc3279#section-2.3.1
+/// [RFC-5280]: https://tools.ietf.org/html/rfc5280#section-4.1
+const ASN1_RSA_ENCRYPTION_OID: [u8;9] = [ 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01 ];
 
 /// The ASN.1 OID for "rsaEncryption".
 #[derive(Clone)]
 struct Asn1OidRsaEncryption();
 
-impl IntoDerObject for Asn1OidRsaEncryption {
-    fn into_der_object(self) -> DerObject {
-        OID_RSA_ENCRYPTION_DER.clone()
-    }
-    fn serialized_len(&self) -> usize {
-        OID_RSA_ENCRYPTION_DER.serialized_len()
+impl asn1_der::typed::DerEncodable for Asn1OidRsaEncryption {
+    fn encode<S: asn1_der::Sink>(&self, sink: &mut S) -> Result<(), asn1_der::Asn1DerError> {
+        // TODO: `DerObject::write` is a hidden method. Should one use `new` or `new_from_source`?
+        // If so, I don't see a way to do that given that `S: Into<&'a [u8]` would be required.
+        asn1_der::DerObject::write(
```

Comment:
Is this the way to go?

Reply:
Yes, the use of write is correct when serializing a new type; I use it for the same purpose and it should not be hidden 🙈

However when defining a new type you should put the implementation into a `DerTypeView`; e.g.:
```rust
/// A raw OID
#[derive(Copy, Clone)]
struct Asn1RawOid<'a> {
    object: DerObject<'a>
}
impl<'a> Asn1RawOid<'a> {
    /// Wraps an OID byte literal
    pub fn new<S: Sink + Into<&'a[u8]>>(oid: &[u8], sink: S) -> Result<Self, asn1_der::Asn1DerError> {
        let object = DerObject::new(Self::TAG, oid, sink)?;
        Ok(Self { object })
    }

    /// The underlying OID as byte literal
    pub fn oid(&self) -> &[u8] {
        self.object.value()
    }

    /// Writes an OID raw `value` as DER-object to `sink`
    pub fn write<S: Sink>(value: &[u8], sink: &mut S) -> Result<(), Asn1DerError> {
        DerObject::write(Self::TAG, value.len(), &mut value.iter(), sink)
    }
}
impl<'a> asn1_der::typed::DerTypeView<'a> for Asn1RawOid<'a> {
    const TAG: u8 = 6;

    fn object(&self) -> DerObject<'a> {
        self.object
    }
}
impl<'a> asn1_der::typed::DerEncodable for Asn1RawOid<'a> {
    fn encode<S: asn1_der::Sink>(&self, sink: &mut S) -> Result<(), asn1_der::Asn1DerError> {
        self.object.encode(sink)
    }
}
impl<'a> asn1_der::typed::DerDecodable<'a> for Asn1RawOid<'a> {
    fn load(object: DerObject<'a>) -> Result<Self, Asn1DerError> {
        // Validate the tag
        if object.tag() != Self::TAG {
            return Err(Asn1DerError::new(
                Asn1DerErrorVariant::InvalidData("DER object tag is not the object identifier tag."),
            ));
        }

        // Load self
        Ok(Self { object })
    }
}
```
(This is because the general approach of this crate is view-based; i.e. the core implementations do not own any bytes).


Now you can easily implement a native type upon it:
```rust
/// An 'rsaEncryption' OID
#[derive(Copy, Clone)]
struct Asn1OidRsaEncryption;
impl Asn1OidRsaEncryption {
    /// The DER encoding of the object identifier (OID) 'rsaEncryption' for
    /// RSA public keys defined for X.509 in [RFC-3279] and used in
    /// SubjectPublicKeyInfo structures defined in [RFC-5280].
    ///
    /// [RFC-3279]: https://tools.ietf.org/html/rfc3279#section-2.3.1
    /// [RFC-5280]: https://tools.ietf.org/html/rfc5280#section-4.1
    const OID: &'static[u8] = b"\x2A\x86\x48\x86\xF7\x0D\x01\x01\x01";
}
impl<'a> DerDecodable<'a> for Asn1OidRsaEncryption {
    fn load(object: DerObject<'a>) -> Result<Self, Asn1DerError> {
        match Asn1RawOid::load(object)?.oid() {
            Self::OID => Ok(Self),
            _ => Err(Asn1DerError::new(
                Asn1DerErrorVariant::InvalidData("DER object is not the 'rsaEncryption' identifier."),
            ))
        }
    }
}
impl DerEncodable for Asn1OidRsaEncryption {
    fn encode<S: Sink>(&self, sink: &mut S) -> Result<(), Asn1DerError> {
        Asn1RawOid::write(Self::OID, sink)
    }
}
```

If it makes things easier for you, I can also add a raw OID type to the core crate; I'm still a bit reluctant about something that *understands* OIDs, but an adapter to work with the raw bytes is no problem anymore IMO.